File: src/Shop.Shared/Diagnostics.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using Microsoft.JSInterop;

namespace Shop.Shared;

/// <summary>Группа ресурсов страницы: рантайм .NET, пакет UI-библиотеки, шрифты и т. д.</summary>
public class ResourceGroupInfo
{
    public string Name { get; set; } = "";
    public int Count { get; set; }
    public long TransferBytes { get; set; }
    public long DecodedBytes { get; set; }
    public int CachedCount { get; set; }
    public double DurationMs { get; set; }
}

/// <summary>
/// Снимок нагрузки приложения, собираемый одинаковым скриптом diagnostics.js во всех фронтендах:
/// тайминги загрузки, отрисовка, состав и вес ресурсов, размер DOM и CSS, JS-куча, устройство.
/// </summary>
public class DiagnosticsReport
{
    // ----- Загрузка страницы -----
    public string NavigationType { get; set; } = "";
    public int RedirectCount { get; set; }
    public double? DnsMs { get; set; }
    public double? TcpMs { get; set; }
    public double? TlsMs { get; set; }
    public double? TtfbMs { get; set; }
    public double? ResponseMs { get; set; }
    public double? DomInteractiveMs { get; set; }
    public double? DomContentLoadedMs { get; set; }
    public double? LoadEventMs { get; set; }
    public long DocumentTransferBytes { get; set; }
    public long DocumentDecodedBytes { get; set; }

    // ----- Отрисовка -----
    public double? FirstPaintMs { get; set; }
    public double? FirstContentfulPaintMs { get; set; }
    public double? LargestContentfulPaintMs { get; set; }
    /// <summary>Момент первой готовой отрисовки интерфейса приложения (MainLayout, firstRender).</summary>
    public double? AppReadyMs { get; set; }
    public int AppRenders { get; set; }
    public int LongTaskCount { get; set; }
    public double? LongTaskTotalMs { get; set; }
    public double? LongestTaskMs { get; set; }

    // ----- Ресурсы -----
    public int ResourceCount { get; set; }
    public long ResourceTransferBytes { get; set; }
    public long ResourceDecodedBytes { get; set; }
    public int ResourceCachedCount { get; set; }
    public double? SlowestResourceMs { get; set; }
    public string SlowestResourceUrl { get; set; } = "";
    public List<ResourceGroupInfo> ResourceGroups { get; set; } = [];

    // ----- DOM и CSS -----
    public int DomNodes { get; set; }
    public int DomDepth { get; set; }
    public int StyleSheetCount { get; set; }
    public int CssRuleCount { get; set; }
    public int UnreadableStyleSheets { get; set; }

    // ----- Память -----
    public long? UsedJsHeapBytes { get; set; }
    public long? TotalJsHeapBytes { get; set; }
    public long? JsHeapLimitBytes { get; set; }

    // ----- Устройство и окружение -----
    public double DevicePixelRatio { get; set; }
    public int ViewportWidth { get; set; }
    public int ViewportHeight { get; set; }
    public int ScreenWidth { get; set; }
    public int ScreenHeight { get; set; }
    public int HardwareConcurrency { get; set; }
    public double DeviceMemoryGb { get; set; }
    public string ConnectionType { get; set; } = "";
    public double ConnectionDownlinkMbps { get; set; }
    public double ConnectionRttMs { get; set; }
    public string Language { get; set; } = "";
    public string UserAgent { get; set; } = "";
    public string CollectedAt { get; set; } = "";

    /// <summary>Имя приложения бенчмарка — подставляет фронтенд, скрипт про него не знает.</summary>
    public string AppName { get; set; } = "";

    public static string Bytes(long value) => value switch
    {
        >= 1_048_576 => $"{value / 1_048_576.0:F2} МБ",
        >= 1024 => $"{value / 1024.0:F1} КБ",
        _ => $"{value} Б",
    };

    public static string Ms(double? value) =>
        value is null ? "нет данных" : $"{value.Value.ToString("N0", CultureInfo.InvariantCulture)} мс";

    private static string Num(double value) => value.ToString("N0", CultureInfo.InvariantCulture);

    /// <summary>
    /// Разделы отчёта — одинаковые во всех трёх приложениях: каждая страница «Диагностика»
    /// только рисует их своими компонентами, набор и порядок показателей общий.
    /// </summary>
    public IReadOnlyList<DiagnosticsSection> ToSections() =>
    [
        new("Отрисовка", [
            new("Первая отрисовка (FP)", Ms(FirstPaintMs)),
            new("Первый контент (FCP)", Ms(FirstContentfulPaintMs)),
            new("Крупнейший элемент (LCP)", Ms(LargestContentfulPaintMs)),
            new("Интерфейс приложения готов", Ms(AppReadyMs)),
            new("Длинных задач (> 50 мс)", $"{LongTaskCount} шт., суммарно {Ms(LongTaskTotalMs)}"),
            new("Самая длинная задача", Ms(LongestTaskMs)),
        ]),
        new("Загрузка страницы", [
            new("Тип навигации", string.IsNullOrEmpty(NavigationType) ? "—" : NavigationType),
            new("DNS / TCP / TLS", $"{Ms(DnsMs)} / {Ms(TcpMs)} / {(TlsMs is null ? "без TLS" : Ms(TlsMs))}"),
            new("Ответ сервера (TTFB)", Ms(TtfbMs)),
            new("Загрузка документа", $"{Ms(ResponseMs)}, {Bytes(DocumentTransferBytes)} (распаковано {Bytes(DocumentDecodedBytes)})"),
            new("DOM interactive", Ms(DomInteractiveMs)),
            new("DOMContentLoaded", Ms(DomContentLoadedMs)),
            new("Load event", Ms(LoadEventMs)),
        ]),
        new("Ресурсы", [
            new("Всего запросов", $"{ResourceCount} шт., из них из кэша {ResourceCachedCount}"),
            new("Передано по сети", Bytes(ResourceTransferBytes)),
            new("Распаковано", Bytes(ResourceDecodedBytes)),
            new("Самый долгий ресурс", SlowestResourceMs is null
                ? "нет данных"
                : $"{Ms(SlowestResourceMs)} — {SlowestResourceUrl}"),
        ]),
        new("DOM и стили", [
            new("Узлов в DOM", Num(DomNodes)),
            new("Глубина DOM", Num(DomDepth)),
            new("Таблиц стилей", Num(StyleSheetCount) + (UnreadableStyleSheets > 0 ? $" (недоступно для чтения: {UnreadableStyleSheets})" : "")),
            new("CSS-правил", CssRuleCount > 0 ? Num(CssRuleCount) : "нет данных"),
        ]),
        new("Память", [
            new("JS-куча, занято", UsedJsHeapBytes is { } used ? Bytes(used) : "нет данных (только Chromium)"),
            new("JS-куча, выделено", TotalJsHeapBytes is { } total ? Bytes(total) : "нет данных"),
            new("JS-куча, предел", JsHeapLimitBytes is { } limit ? Bytes(limit) : "нет данных"),
        ]),
        new("Устройство", [
            new("Окно / экран", $"{ViewportWidth}×{ViewportHeight} / {ScreenWidth}×{ScreenHeight}"),
            new("Device pixel ratio", DevicePixelRatio.ToString("0.##", CultureInfo.InvariantCulture)),
            new("Ядер CPU / память", $"{(HardwareConcurrency > 0 ? HardwareConcurrency.ToString() : "—")} / {(DeviceMemoryGb > 0 ? DeviceMemoryGb + " ГБ" : "—")}"),
            new("Сеть", string.IsNullOrEmpty(ConnectionType)
                ? "нет данных"
                : $"{ConnectionType}, {ConnectionDownlinkMbps} Мбит/с, RTT {ConnectionRttMs} мс"),
            new("Язык браузера", string.IsNullOrEmpty(Language) ? "—" : Language),
            new("User agent", UserAgent),
        ]),
    ];

    /// <summary>Текстовый отчёт для буфера обмена: удобно складывать рядом три приложения.</summary>
    public string ToPlainText()
    {
        var text = new StringBuilder();
        text.AppendLine($"Диагностика — {(string.IsNullOrEmpty(AppName) ? "приложение" : AppName)} ({CollectedAt})");
        foreach (var section in ToSections())
        {
            text.AppendLine();
            text.AppendLine($"[{section.Title}]");
            foreach (var row in section.Rows)
                text.AppendLine($"{row.Name}: {row.Value}");
        }

        text.AppendLine();
        text.AppendLine("[Ресурсы по группам]");
        foreach (var group in ResourceGroups)
            text.AppendLine($"{group.Name}: {group.Count} шт., {Bytes(group.TransferBytes)} по сети, {Bytes(group.DecodedBytes)} распаковано");
        return text.ToString();
    }

    public string ToJson() => JsonSerializer.Serialize(this, DiagnosticsJson.Options);
}

/// <summary>Раздел отчёта диагностики.</summary>
public sealed record DiagnosticsSection(string Title, IReadOnlyList<DiagnosticsRow> Rows);

/// <summary>Строка отчёта диагностики: показатель и значение.</summary>
public sealed record DiagnosticsRow(string Name, string Value);

internal static class DiagnosticsJson
{
    public static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };
}

/// <summary>
/// Служба диагностики: читает метрики страницы через идентичный diagnostics.js,
/// подключённый в каждом из трёх приложений. Один код — одинаковые показатели.
/// </summary>
public class DiagnosticsService(IJSRuntime js)
{
    /// <summary>Отмечает первую готовую отрисовку интерфейса; вызывается из MainLayout (firstRender).</summary>
    public async Task MarkAppReadyAsync()
    {
        try { await js.InvokeVoidAsync("shopDiagnostics.markAppReady"); }
        catch (JSDisconnectedException) { }
        catch (JSException) { }
        catch (InvalidOperationException) { }
    }

    public async Task<DiagnosticsReport?> CollectAsync(string appName = "")
    {
        try
        {
            var report = await js.InvokeAsync<DiagnosticsReport>("shopDiagnostics.collect");
            if (report is not null) report.AppName = appName;
            return report;
        }
        catch (JSDisconnectedException) { return null; }
        catch (JSException) { return null; }
        catch (InvalidOperationException) { return null; }
    }

    /// <summary>Копирует текстовый отчёт в буфер обмена. Возвращает false, если браузер не дал доступ.</summary>
    public async Task<bool> CopyAsync(string text)
    {
        try { return await js.InvokeAsync<bool>("shopDiagnostics.copyReport", text); }
        catch (JSDisconnectedException) { return false; }
        catch (JSException) { return false; }
        catch (InvalidOperationException) { return false; }
    }
}

File: src/Shop.Shared/wwwroot/diagnostics.js
// Идентичный скрипт диагностики для всех трёх фронтендов (Flare / MudBlazor / Radzen).
// Подключается в <head> до blazor.webassembly.js, поэтому успевает подписаться на события
// отрисовки и длинные задачи ещё до старта .NET — иначе FCP/LCP и longtask уже не собрать.
//
// Собирает всё, чем приложения отличаются друг от друга при одинаковом коде страниц:
// тайминги загрузки, вес и состав ресурсов (рантайм Blazor / пакет библиотеки / шрифты),
// объём CSS темы, размер DOM, JS-кучу и параметры устройства.
(function () {
    'use strict';

    var observed = {
        firstPaintMs: null,
        firstContentfulPaintMs: null,
        largestContentfulPaintMs: null,
        longTaskCount: 0,
        longTaskTotalMs: 0,
        longestTaskMs: 0,
        appReadyMs: null,
        appRenders: 0
    };

    function observe(type, handler, extra) {
        try {
            var options = { type: type, buffered: true };
            if (extra) for (var key in extra) options[key] = extra[key];
            new PerformanceObserver(function (list) {
                var entries = list.getEntries();
                for (var i = 0; i < entries.length; i++) handler(entries[i]);
            }).observe(options);
        } catch (e) { /* тип не поддерживается этим браузером — метрика останется null */ }
    }

    observe('paint', function (entry) {
        if (entry.name === 'first-paint') observed.firstPaintMs = entry.startTime;
        if (entry.name === 'first-contentful-paint') observed.firstContentfulPaintMs = entry.startTime;
    });

    observe('largest-contentful-paint', function (entry) {
        observed.largestContentfulPaintMs = entry.startTime;
    });

    observe('longtask', function (entry) {
        observed.longTaskCount++;
        observed.longTaskTotalMs += entry.duration;
        if (entry.duration > observed.longestTaskMs) observed.longestTaskMs = entry.duration;
    });

    function round(value) {
        return (value === null || value === undefined || isNaN(value)) ? null : Math.round(value);
    }

    // Ресурс относится к одной из групп сравнения: рантайм Blazor, пакет UI-библиотеки,
    // шрифты, собственные файлы приложения. Имя пакета берём прямо из пути _content/<пакет>/.
    function groupOf(entry) {
        var url = entry.name;
        if (url.indexOf('/_framework/') >= 0) {
            if (/dotnet[.\-][^/]*\.js(\?|$)/.test(url) || /dotnet\.native/.test(url) || /dotnet\.runtime/.test(url))
                return 'Рантайм .NET';
            if (/\.(wasm|dll)(\?|$)/.test(url)) return 'Сборки приложения (.NET)';
            return 'Загрузчик Blazor';
        }
        var content = url.match(/\/_content\/([^/]+)\//);
        if (content) return 'Пакет ' + content[1];
        if (/fonts\.(googleapis|gstatic)\.com/.test(url) || /\.(woff2?|ttf|otf|eot)(\?|$)/.test(url)) return 'Шрифты';
        if (entry.initiatorType === 'css' || /\.css(\?|$)/.test(url)) return 'CSS приложения';
        if (entry.initiatorType === 'img' || /\.(png|jpe?g|gif|svg|webp|ico)(\?|$)/.test(url)) return 'Изображения';
        return 'Прочее';
    }

    function collectResources() {
        var entries = performance.getEntriesByType('resource');
        var groups = {};
        var totals = { count: 0, transferBytes: 0, decodedBytes: 0, cachedCount: 0, slowestMs: 0, slowestUrl: '' };

        for (var i = 0; i < entries.length; i++) {
            var entry = entries[i];
            var name = groupOf(entry);
            var group = groups[name] || (groups[name] = {
                name: name, count: 0, transferBytes: 0, decodedBytes: 0, cachedCount: 0, durationMs: 0
            });

            var transfer = entry.transferSize || 0;
            var decoded = entry.decodedBodySize || 0;

            group.count++;
            group.transferBytes += transfer;
            group.decodedBytes += decoded;
            group.durationMs += entry.duration || 0;
            if (transfer === 0 && decoded > 0) group.cachedCount++;

            totals.count++;
            totals.transferBytes += transfer;
            totals.decodedBytes += decoded;
            if (transfer === 0 && decoded > 0) totals.cachedCount++;
            if ((entry.duration || 0) > totals.slowestMs) {
                totals.slowestMs = entry.duration;
                totals.slowestUrl = entry.name.replace(/^https?:\/\/[^/]+/, '');
            }
        }

        var list = [];
        for (var key in groups) {
            groups[key].durationMs = round(groups[key].durationMs);
            list.push(groups[key]);
        }
        list.sort(function (a, b) { return b.transferBytes - a.transferBytes || b.decodedBytes - a.decodedBytes; });

        totals.slowestMs = round(totals.slowestMs);
        return { totals: totals, groups: list };
    }

    // Вес темы: сколько CSS-правил и селекторов реально подключила библиотека.
    // Кросс-доменные таблицы (шрифты Google) правила не отдают — они считаются как недоступные.
    function collectCss() {
        var sheets = document.styleSheets;
        var rules = 0, unreadable = 0;
        for (var i = 0; i < sheets.length; i++) {
            try {
                var sheetRules = sheets[i].cssRules;
                rules += sheetRules ? sheetRules.length : 0;
            } catch (e) { unreadable++; }
        }
        return { styleSheetCount: sheets.length, cssRuleCount: rules, unreadableStyleSheets: unreadable };
    }

    function maxDomDepth() {
        var max = 0;
        var walk = function (node, depth) {
            if (depth > max) max = depth;
            var children = node.children;
            for (var i = 0; i < children.length; i++) walk(children[i], depth + 1);
        };
        walk(document.documentElement, 1);
        return max;
    }

    function collectNavigation() {
        var nav = performance.getEntriesByType('navigation')[0];
        if (!nav) return null;
        return {
            type: nav.type,
            redirectCount: nav.redirectCount,
            dnsMs: round(nav.domainLookupEnd - nav.domainLookupStart),
            tcpMs: round(nav.connectEnd - nav.connectStart),
            tlsMs: nav.secureConnectionStart ? round(nav.connectEnd - nav.secureConnectionStart) : null,
            ttfbMs: round(nav.responseStart - nav.requestStart),
            responseMs: round(nav.responseEnd - nav.responseStart),
            domInteractiveMs: round(nav.domInteractive),
            domContentLoadedMs: round(nav.domContentLoadedEventEnd),
            loadEventMs: round(nav.loadEventEnd),
            documentTransferBytes: nav.transferSize || 0,
            documentDecodedBytes: nav.decodedBodySize || 0
        };
    }

    window.shopDiagnostics = {
        // Приложение отмечает момент первой готовой отрисовки (MainLayout, firstRender).
        // Это и есть «сколько ждёт пользователь до рабочего интерфейса» — главный показатель сравнения.
        markAppReady: function () {
            observed.appRenders++;
            if (observed.appReadyMs === null) {
                observed.appReadyMs = performance.now();
                try { performance.mark('shop-app-ready'); } catch (e) { /* нет поддержки User Timing */ }
            }
            return observed.appReadyMs;
        },

        collect: function () {
            var nav = collectNavigation();
            var resources = collectResources();
            var css = collectCss();
            var memory = performance.memory; // Chromium-only; в остальных браузерах undefined
            var connection = navigator.connection || {};

            return {
                // ----- Загрузка страницы -----
                navigationType: nav ? nav.type : '',
                redirectCount: nav ? nav.redirectCount : 0,
                dnsMs: nav ? nav.dnsMs : null,
                tcpMs: nav ? nav.tcpMs : null,
                tlsMs: nav ? nav.tlsMs : null,
                ttfbMs: nav ? nav.ttfbMs : null,
                responseMs: nav ? nav.responseMs : null,
                domInteractiveMs: nav ? nav.domInteractiveMs : null,
                domContentLoadedMs: nav ? nav.domContentLoadedMs : null,
                loadEventMs: nav ? nav.loadEventMs : null,
                documentTransferBytes: nav ? nav.documentTransferBytes : 0,
                documentDecodedBytes: nav ? nav.documentDecodedBytes : 0,

                // ----- Отрисовка -----
                firstPaintMs: round(observed.firstPaintMs),
                firstContentfulPaintMs: round(observed.firstContentfulPaintMs),
                largestContentfulPaintMs: round(observed.largestContentfulPaintMs),
                appReadyMs: round(observed.appReadyMs),
                appRenders: observed.appRenders,
                longTaskCount: observed.longTaskCount,
                longTaskTotalMs: round(observed.longTaskTotalMs),
                longestTaskMs: round(observed.longestTaskMs),

                // ----- Ресурсы -----
                resourceCount: resources.totals.count,
                resourceTransferBytes: resources.totals.transferBytes,
                resourceDecodedBytes: resources.totals.decodedBytes,
                resourceCachedCount: resources.totals.cachedCount,
                slowestResourceMs: resources.totals.slowestMs,
                slowestResourceUrl: resources.totals.slowestUrl,
                resourceGroups: resources.groups,

                // ----- DOM и CSS -----
                domNodes: document.getElementsByTagName('*').length,
                domDepth: maxDomDepth(),
                styleSheetCount: css.styleSheetCount,
                cssRuleCount: css.cssRuleCount,
                unreadableStyleSheets: css.unreadableStyleSheets,

                // ----- Память -----
                usedJsHeapBytes: memory ? memory.usedJSHeapSize : null,
                totalJsHeapBytes: memory ? memory.totalJSHeapSize : null,
                jsHeapLimitBytes: memory ? memory.jsHeapSizeLimit : null,

                // ----- Устройство и окружение -----
                devicePixelRatio: window.devicePixelRatio || 1,
                viewportWidth: window.innerWidth,
                viewportHeight: window.innerHeight,
                screenWidth: screen.width,
                screenHeight: screen.height,
                hardwareConcurrency: navigator.hardwareConcurrency || 0,
                deviceMemoryGb: navigator.deviceMemory || 0,
                connectionType: connection.effectiveType || '',
                connectionDownlinkMbps: connection.downlink || 0,
                connectionRttMs: connection.rtt || 0,
                language: navigator.language || '',
                userAgent: navigator.userAgent,
                collectedAt: new Date().toISOString()
            };
        },

        // Отчёт в буфер обмена — чтобы сравнивать три приложения, а не пересказывать цифры.
        copyReport: function (text) {
            if (!navigator.clipboard) return false;
            navigator.clipboard.writeText(text);
            return true;
        }
    };
})();
